File: src/main/java/sawyern/cookiebot/services/DiscordService.java
package sawyern.cookiebot.services;

import discord4j.core.DiscordClient;
import discord4j.core.DiscordClientBuilder;
import discord4j.core.event.domain.lifecycle.ReadyEvent;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;
import sawyern.cookiebot.commands.BotCommand;
import sawyern.cookiebot.exception.AssertMessage;
import sawyern.cookiebot.properties.DiscordClientProperties;

import java.util.List;

@Service
@Slf4j
@NoArgsConstructor
public class DiscordService {

    @Setter
    private DiscordClientProperties discordClientProperties;
    @Setter
    private List<BotCommand> botCommands;

    @Getter(AccessLevel.PROTECTED)
    private DiscordClient discordClient;

    /**
     * Create discord client given token from properties file
     */
    public void setDiscordClient() {
        if (getDiscordClient() != null)
            return;

        Assert.state(discordClientProperties != null, AssertMessage.ERROR_LOADING_PROPERTY);
        Assert.state(discordClientProperties.getToken() != null, AssertMessage.DISCORD_TOKEN_NOT_SET);
        discordClient = new DiscordClientBuilder(discordClientProperties.getToken()).build();
    }

    /**
     * Subscribe commands to discord events
     */
    public void subscribeAllCommands() {
        Assert.state(getDiscordClient() != null, AssertMessage.DISCORD_CLIENT_NOT_SET);
        Assert.state(botCommands != null, AssertMessage.NULL_BOT_COMMANDS);
        botCommands.forEach(command -> command.subscribeCommand(getDiscordClient()));
    }

    /**
     * log ready event
     */
    public void subscribeReady() {
        Assert.state(getDiscordClient() != null, AssertMessage.DISCORD_CLIENT_NOT_SET);
        getDiscordClient().getEventDispatcher()
                .on(ReadyEvent.class)
                .subscribe(ready -> log.info("Logged in as {}", ready.getSelf().getUsername()));
    }

    /**
     * discord client will log into server
     */
    public void login() {
        Assert.state(getDiscordClient() != null, AssertMessage.DISCORD_CLIENT_NOT_SET);
        getDiscordClient().login().block();
    }
}

File: src/main/kotlin/sawyern/cookiebot/services/DiscordService.kt
package sawyern.cookiebot.services

import discord4j.common.util.Snowflake
import discord4j.core.DiscordClient
import discord4j.core.GatewayDiscordClient
import discord4j.core.`object`.entity.channel.MessageChannel
import discord4j.core.event.domain.lifecycle.ReadyEvent
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import sawyern.cookiebot.commands.BotCommand
import sawyern.cookiebot.properties.DiscordClientProperties

@Service
class DiscordService @Autowired
constructor(
        private val discordClientProperties: DiscordClientProperties,
        private val botCommands: List<BotCommand>
) {
    private val discordClient: DiscordClient by lazy {
        DiscordClient.create(discordClientProperties.token)
    }

    private val logger = KotlinLogging.logger {}

    internal fun subscribeReady(gateway : GatewayDiscordClient) {
        gateway.on(ReadyEvent::class.java).subscribe { logger.info("Logged in as {}", it.self.username) }
    }

    internal fun subscribeAllCommands(gateway : GatewayDiscordClient) {
        botCommands.forEach { command -> command.subscribe(gateway) }
    }

    internal fun login() : GatewayDiscordClient? { return discordClient.login().block() }

    internal fun getBotChannel(): MessageChannel {
        return discordClient
                .getChannelById(Snowflake.of(discordClientProperties.botChannelId))
                as MessageChannel
    }
}




